src/Stratis.SmartContracts.CLR/Caching/RewrittenContractCache.cs

```diff
@@ -12,38 +11,46 @@ public interface IRewrittenContractCache
         ///
         /// If trying to add a cache entry twice for the same hash, this will throw an exception.
         /// </summary>
-        void Store(uint256 initialCodeHash, ModuleDefinition module);
+        void Store(uint256 initialCodeHash, byte[] rewrittenContractCode);
 
         /// <summary>
         /// Returns a cached module definition for this contract code if one exists.
         ///
         /// Returns null if this contract code has not been cached yet.
         /// </summary>
-        ModuleDefinition Retrieve(uint256 codeHash);
+        byte[] Retrieve(uint256 codeHash);
     }
 
     public class RewrittenContractCache : IRewrittenContractCache
     {
-        private readonly Dictionary<uint256, ModuleDefinition> cachedContracts;
+        private readonly Dictionary<uint256, byte[]> cachedContracts;
 
-        public RewrittenContractCache()
+        private readonly ILogger logger;
+
+        public RewrittenContractCache(ILoggerFactory loggerFactory)
         {
-            this.cachedContracts = new Dictionary<uint256, ModuleDefinition>();
+            this.cachedContracts = new Dictionary<uint256, byte[]>();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType());
         }
 
         /// <inheritdoc />
-        public void Store(uint256 initialCodeHash, ModuleDefinition module)
+        public void Store(uint256 initialCodeHash, byte[] rewrittenContractCode)
         {
             if (this.cachedContracts.ContainsKey(initialCodeHash))
-                throw new Exception("Cache already contains an entry for this code hash.");
+            {
+                // This could happen in a race condition where multiple threads are using the same contract for the first time at the same time.
+                this.logger.LogWarning("Tried to add the same contract code to the cache twice.");
+                return;
+            }
```

Comment:
If a race condition is a concern here this won't really protect against it. The contract could be added after this block executes but before the next line executes.

Reply:
That's okay. The logs will tell us this has happened. As long as there is one valid contract in there after it occurs.

If this log is happening a lot, something is wrong. It's not necessarily an error if it only happens occasionally, but if we see this log a lot then we should look into it. I figured `LogWarning` is built for nearly this exact purpose!

Reply:
👍

Reply:
This could still be a problem. If `Add` is called below and the key already exists, an exception will be thrown. We can use `this.cachedContracts[initialCodeHash]` to perform an add or update instead.

Reply:
Good point!